File: diff_benchmark/diff.f90
subroutine init(a, at, itot, jtot, ktot)
    integer, intent(in) :: itot, jtot, ktot
    real*8, dimension(itot,jtot,ktot), intent(inout) :: a, at
    integer :: ijk

    do k=1,ktot
        do j=1,jtot
            do i=1,itot
                ijk = i + (j-1)*itot + (k-1)*itot*jtot
                a(i,j,k)  = (ijk-1)**2. / ijk**2.     
                at(i,j,k) = 0
            end do
        end do
    end do
end subroutine init

subroutine diff(at, a, visc, dxidxi, dyidyi, dzidzi, itot, jtot, ktot)
    integer, intent(in) :: itot, jtot, ktot
    real*8, intent(in) :: visc, dxidxi, dyidyi, dzidzi
    real*8, dimension(itot,jtot,ktot), intent(inout) :: at
    real*8, dimension(itot,jtot,ktot), intent(in) :: a

    do k=2,ktot-1
        do j=2,jtot-1
            do i=2,itot-1
                at(i,j,k) = at(i,j,k) + visc * ( &
                        + ( (a(i+1,j,k) - a(i,j,k  )) &
                          - (a(i,j,k  ) - a(i-1,j,k)) ) * dxidxi &
                        + ( (a(i,j+1,k) - a(i,j,k  )) &
                          - (a(i,j,k  ) - a(i,j-1,k)) ) * dyidyi &
                        + ( (a(1,j,k+1) - a(i,j,k  )) &
                          - (a(i,j,k  ) - a(i,j,k-1)) ) * dzidzi &
                        )
            end do
        end do
    end do
end subroutine diff

program main
    integer :: nloop = 100
    integer :: itot  = 128
    integer :: jtot  = 128
    integer :: ktot  = 128

    real :: starttime, endtime
    real*8 :: tmp = 0.1

    real*8, dimension(:,:,:), allocatable :: a, at
    allocate(a(itot,jtot,ktot), at(itot,jtot,ktot))

    call init(a, at, itot, jtot, ktot)

    ! Check results
    call diff(at, a, tmp, tmp, tmp, tmp, itot, jtot, ktot)
    print*,"at=",at(itot/2+1, 2, 2)

    ! Time performance 
    call cpu_time(starttime)

    do i=1, nloop
        call diff(at, a, tmp, tmp, tmp, tmp, itot, jtot, ktot)
    end do

    call cpu_time(endtime)
    print*, "time/iter =", (endtime-starttime)/nloop, " s (", nloop, " iters)"
end program main

File: diff_benchmark/diff.cxx
#include <iostream>
#include <iomanip>
#include <cstdlib>
#include <stdlib.h>
#include <cstdio>
#include <ctime>
#include "math.h"

void init(double* const __restrict__ a, double* const __restrict__ at, const int ncells)
{
    for (int i=0; i<ncells; ++i)
    {
        a[i]  = pow(i,2)/pow(i+1,2);
        at[i] = 0.;
    }
}

void diff(double* const __restrict__ at, const double* const __restrict__ a, const double visc, 
          const double dxidxi, const double dyidyi, const double dzidzi, 
          const int itot, const int jtot, const int ktot)
{
    const int ii = 1;
    const int jj = itot;
    const int kk = itot*jtot;

    for (int k=1; k<ktot-1; k++)
        for (int j=1; j<jtot-1; j++)
        #pragma ivdep
            for (int i=1; i<itot-1; i++)
            {
                const int ijk = i + j*jj + k*kk;
                at[ijk] += visc * (
                        + ( (a[ijk+ii] - a[ijk   ]) 
                          - (a[ijk   ] - a[ijk-ii]) ) * dxidxi 
                        + ( (a[ijk+jj] - a[ijk   ]) 
                          - (a[ijk   ] - a[ijk-jj]) ) * dyidyi
                        + ( (a[ijk+kk] - a[ijk   ]) 
                          - (a[ijk   ] - a[ijk-kk]) ) * dzidzi
                        );
            }
}

int main()
{
    const int nloop = 100;
    const int itot = 128;
    const int jtot = 128;
    const int ktot = 128;
    const int ncells = itot*jtot*ktot;

    double *a  = new double[ncells];
    double *at = new double[ncells];
   
    init(a, at, ncells);

    // Check results
    diff(at, a, 0.1, 0.1, 0.1, 0.1, itot, jtot, ktot); 
    printf("at=%.20f\n",at[itot*jtot+itot+itot/2]);
 
    // Time performance 
    std::clock_t start = std::clock(); 
   
    for (int i=0; i<nloop; ++i)
        diff(at, a, 0.1, 0.1, 0.1, 0.1, itot, jtot, ktot); 
  
    double duration = (std::clock() - start ) / (double)CLOCKS_PER_SEC;
   
    printf("time/iter = %f s (%i iters)\n",duration/(double)nloop, nloop);
    
    return 0;
}
